emukit/core/optimization/local_search_acquisition_optimizer.py

```diff
@@ -0,0 +1,176 @@
+        """ Generates one-exchange neighbours of encoded parameter point.
+
+        :param all_features: The encoded parameter point (1d-array)
+        :return: All one-exchange neighbours as 2d-array (neighbours, features)
+        """
+        neighbours_per_param = self._neighbours_per_parameter(
+            all_features, parameters)
+        num_neighbours = sum(param.shape[0] for param in neighbours_per_param)
+        neighbours = np.full((num_neighbours, all_features.shape[0]), all_features)
+        current_neighbour, current_feature = 0, 0
+        for this_neighbours in neighbours_per_param:
+            next_neighbour = current_neighbour + this_neighbours.shape[0]
+            next_feature = current_feature + this_neighbours.shape[1]
+            neighbours[current_neighbour:next_neighbour,
+                       current_feature:next_feature] = this_neighbours
+            current_neighbour, current_feature = next_neighbour, next_feature
+        return neighbours
+
+    def _one_local_search(self, acquisition: Acquisition, x: np.ndarray,
+                          noncontext_space: ParameterSpace,
+                          context_manager: Optional[ContextManager] = None):
+        """ Local maximum search on acquisition starting at a single point.
+
+        :param acquisition: The acquisition the maximum is searched of.
+        :param x: The initial point.
+        :return: Tuple of (maximum point as 1d-array, value of acquisition at this point)
+        """
+        incumbent_value = acquisition.evaluate(x.reshape(1, -1)).item()
+        _log.debug("Start local search with acquisition={:.4f} at {}"
+                   .format(incumbent_value, str(x)))
+        for step in range(self.num_steps):
+            neighbours = self._neighbours(x, noncontext_space.parameters)
+            if context_manager is not None:
+                neighbours_with_context = context_manager._expand_vector(neighbours)
+            else:
+                neighbours_with_context = neighbours
+            acquisition_values = acquisition.evaluate(neighbours_with_context)
+            max_index = np.argmax(acquisition_values)
+            max_neighbour = neighbours[max_index]
+            max_value = np.asscalar(acquisition_values[max_index])
+            if max_value < incumbent_value:
+                _log.debug("End after {} steps at maximum of acquisition={:.4f} at {}"
+                           .format(step, incumbent_value, str(x)))
+                return x, incumbent_value
+            else:
+                incumbent_value = max_value
+                x = max_neighbour
+        _log.debug("End at step limit with acquisition={:.4f} at {}"
+                   .format(incumbent_value, str(x)))
+        return x, incumbent_value
+
+    def optimize(self, acquisition: Acquisition, context: dict = None) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Optimizes the acquisition function, taking into account gradients if it supports them
```

Comment:
I don't think this one actually takes gradients into account

Reply:
True, this is a c/p mistake.